File: Core Service/src/main/java/org/hust/audioSearch/db/Mongo.java
package org.hust.audioSearch.db;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.UpdateOptions;
import com.mongodb.client.model.Updates;
import com.mongodb.client.result.UpdateResult;
import org.bson.Document;
import org.hust.audioSearch.shazam.Couple;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Mongo {
    MongoClient mongoClient;
    MongoDatabase mongoDatabase;
    String collectionName = "AddressHashTable";

    public Mongo(String uri, String database) {
        mongoClient = MongoClients.create(uri);
        mongoDatabase = mongoClient.getDatabase(database);
    }

    public void insertFingerprints(Map<Integer, ArrayList<Couple>> fingerprints) {
        MongoCollection<Document> collection = mongoDatabase.getCollection(collectionName);
        for (Map.Entry<Integer, ArrayList<Couple>> entry : fingerprints.entrySet()) {
            List<Document> coupleDocs = new ArrayList<>();
            for (Couple couple : entry.getValue()) {
                Document coupleDoc  = new Document("SongId", couple.songId).append("Timestamp", couple.time);
                coupleDocs.add(coupleDoc);
            }
            UpdateResult result = collection.updateOne(
                    Filters.eq("_id", entry.getKey()),
                    Updates.addEachToSet("Couples", coupleDocs),
                    new UpdateOptions().upsert(true)
            );
            if (result.getUpsertedId()==null&&result.getMatchedCount()==0) System.out.printf("Unable to update hash %s \n", entry.getKey());
        }
    }

    public Map<Integer,ArrayList<Couple>> getMatchingCouples( Integer[] hashes){
        MongoCollection<Document> collection = mongoDatabase.getCollection(collectionName);
        Map<Integer,ArrayList<Couple>> matchedCouples = new HashMap<>() ;
        for (int hash: hashes){
            Document matchdoc = collection.find(new Document("_id", hash)).first();
            if (matchdoc!=null){
                ArrayList<Couple> couplesForAHash = new ArrayList<>();
                ArrayList<Document> coupledocs = (ArrayList<Document>) matchdoc.get("Couples");
                for (Document coupledoc : coupledocs){
                    couplesForAHash.add(new Couple((Integer) coupledoc.get("Timestamp"), (int)coupledoc.get("SongId")));
                }
                matchedCouples.put(hash,couplesForAHash);
            }
//            else System.out.println(String.format("Can't get matched couples for hash %s", hash ));
        }
        return matchedCouples;
    }

    public void close(){
        this.mongoClient.close();
    }
}

File: FileManagementService/audioRecognition/db/Mongo.js
const mongoose = require("mongoose");
const { Schema, model } = mongoose;

// Define the Couple schema
const coupleSchema = new Schema({
  SongId: Number,
  Timestamp: Number,
}, {_id : false});

const Couple = model("Couple", coupleSchema);

const fingerprintSchema = new Schema({
    _id: Number,
    Couples: [coupleSchema]
}, { versionKey: false })

const Fingerprint = model("Fingerprint", fingerprintSchema);

// Define the main MongoDB class
class Mongo {
  constructor(uri) {
    this.uri = uri;
    this.collectionName = "Fingerprints";
    this.connection = null;
  }

  // Establish connection to MongoDB
  async connect() {
    try {
      this.connection = await mongoose.connect(this.uri);
      console.log("Connected to MongoDB");
    } catch (err) {
      console.error("Error connecting to MongoDB:", err);
      throw err;
    }
  }

  // Insert fingerprints into MongoDB
  async insertFingerprints(fingerprints) {
    for (const [hash, couples] of fingerprints.entries()) {
        const coupleDocs = couples.map(couple => ({
        SongId: couple.songId,
        Timestamp: couple.time
      })); 
      const result = await Fingerprint.updateOne(
        { _id: parseInt(hash) },
        {
          $addToSet: { Couples: { $each: coupleDocs } },
        },
        { upsert: true }
      );
      if (result.upsertedCount === 0 && result.matchedCount === 0) {
        console.log(`Unable to update hash ${hash}`);
      }
    }
  }

  // Get matching couples from MongoDB
  async getMatchingCouples(hashes) {
    const collection = mongoose.connection.collection(this.collectionName);
    const matchedCouples = {};

    for (const hash of hashes) {
      const matchDoc = await collection.findOne({ _id: hash });
      if (matchDoc) {
        const couplesForAHash = matchDoc.Couples.map((coupleDoc) => {
          return new Couple({
            SongId: coupleDoc.SongId,
            Timestamp: coupleDoc.Timestamp,
          });
        });
        matchedCouples[hash] = couplesForAHash;
      }
    }

    return matchedCouples;
  }

  async getAllDocuments(collectionName) {
    try {
      const collection = this.connection.connection.collection(collectionName);
      const documents = await collection.find().toArray();
      return documents;
    } catch (err) {
      console.error("Error fetching documents:", err);
      throw err;
    }
  }

  // Close the MongoDB connection
  async close() {
    try {
      await mongoose.connection.close();
      console.log("Disconnected from MongoDB");
    } catch (err) {
      console.error("Error disconnecting from MongoDB:", err);
    }
  }
}

module.exports = { Mongo };
